Homework #2

In [2]:
import numpy as np
from scipy.stats import binom

In [13]:
## Call Payoff Function
def callPayoff(spot, strike):
    return np.maximum(spot - strike, 0.0)

## Put Payoff Function
def putPayoff(spot, strike):
    return np.maximum(strike - spot, 0.0)

In [17]:
def euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, payoff, verbose = True):
    nodes = n  + 1
    h = T / n
    u = 1.3
    d = 0.8
    pu = (np.exp((r - q) * h) - d) / (u - d)
    pd = 1.0 - pu
    disc = np.exp(-r * h)
    
    ## Arrays to store the spot prices and option values
    Ct = np.zeros((nodes, n+1))
    St = np.zeros((nodes, n+1))
    Dt = np.zeros((nodes, n+1))
    Bt = np.zeros((nodes, n+1))
    
    for i in range(nodes):
        St[i, n] = S * (u**(n-i)) * (d**i)
        Ct[i, n] = payoff(St[i, n], K)
    

    for t in range((n-1), -1, -1):
        for j in range(t+1):
            St[j, t] = St[j, t+1] / u
            Ct[j, t] = disc * ((pu * Ct[j, t+1]) + (pd * Ct[j+1, t+1]))
            delta = Dt[j, t] = np.exp(-q * h) * (Ct[j, t+1] - Ct[j+1, t+1]) / (St[j, t] * (u - d))
            beta = Bt[j, t] = np.exp(-r * h) * (((u * Ct[j + 1, t+1]) - (d * Ct[j, t+ 1]))/(u - d))
            print(f"The beta is: {beta : 0.3f} and the delta is {delta : 0.3f}")
            
    if verbose:
        print(np.around(St, 3))
        print("\n")
        print(np.around(Ct, 3))
        print("\n")
        print(np.around(Dt, 3))
        print("\n")
        print(np.around(Bt, 3))         
        print("\n")
        
    return Ct[0,0]

In [32]:
def americanBinomPricerRecursive(S, K, r, v, q, T, n, payoff, verbose = True):
    nodes = n  + 1
    h = T / n
    u = np.exp((r - q) * h + v * np.sqrt(h))
    d = np.exp((r - q) * h - v * np.sqrt(h))
    pu = (np.exp((r - q) * h) - d) / (u - d)
    pd = 1.0 - pu
    disc = np.exp(-r * h)
    
    
    ## Arrays to store the spot prices and option values
    Ct = np.empty(nodes)
    St = np.empty(nodes)
    
    for i in range(nodes):
        St[i] = S * (u ** (n - i)) * (d ** i)
        Ct[i] = payoff(St[i], K)
    
    if verbose:
        print(Ct)
        
    for t in range((n - 1), -1, -1):
        for j in range(t+1):
            Ct[j] = disc * (pu * Ct[j] + pd * Ct[j+1])
            St[j] = St[j] / u
            Ct[j] = np.maximum(Ct[j], payoff(St[j], K))
            print(Ct)
            
    return Ct[0]

Problem #1

In [19]:
S = 100
K = 105
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 1

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, callPayoff,  verbose = False)
putPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, putPayoff,  verbose = False)
print(f"The European Call Premium: {callPrc : 0.3f} and the Put Premium is {putPrc : 0.3f}")

The beta is: -38.432 and the delta is  0.500
The beta is:  62.451 and the delta is -0.500
The European Call Premium:  11.568 and the Put Premium is  12.451


Problem #2

In [21]:
S = 100
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 1

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, callPayoff,  verbose = False)
putPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, putPayoff,  verbose = False)
print(f"The European Call Premium: {callPrc : 0.3f} and the Put Premium is {putPrc : 0.3f}")

The beta is: -53.804 and the delta is  0.700
The beta is:  37.471 and the delta is -0.300
The European Call Premium:  16.196 and the Put Premium is  7.471


Should long the call for the first on and short the second one. The payoff is $0.81 and $0.69

In [24]:
S = 100
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 1

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, callPayoff,  verbose = True)
print(f"The European Call Premium: {callPrc : 0.3f}")

The beta is: -53.804 and the delta is  0.700
[[100. 130.]
 [  0.  80.]]


[[16.196 35.   ]
 [ 0.     0.   ]]


[[0.7 0. ]
 [0.  0. ]]


[[-53.804   0.   ]
 [  0.      0.   ]]


The European Call Premium:  16.196


Problem #4

In [25]:
S = 80
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 3

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, callPayoff,  verbose = False)
print(f"The European Call Premium: {callPrc : 0.3f}")

The beta is: -93.742 and the delta is  1.000
The beta is: -20.777 and the delta is  0.316
The beta is:  0.000 and the delta is  0.000
The beta is: -51.234 and the delta is  0.691
The beta is: -8.751 and the delta is  0.173
The beta is: -26.529 and the delta is  0.457
The European Call Premium:  9.996


In [26]:
S = 90
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 3

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, callPayoff,  verbose = False)
print(f"The European Call Premium: {callPrc : 0.3f}")

The beta is: -93.742 and the delta is  1.000
The beta is: -42.123 and the delta is  0.570
The beta is:  0.000 and the delta is  0.000
The beta is: -63.306 and the delta is  0.805
The beta is: -17.742 and the delta is  0.312
The beta is: -36.698 and the delta is  0.582
The European Call Premium:  15.707


In [27]:
S = 110
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 3

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, callPayoff,  verbose = False)
print(f"The European Call Premium: {callPrc : 0.3f}")

The beta is: -93.742 and the delta is  1.000
The beta is: -84.814 and the delta is  0.939
The beta is:  0.000 and the delta is  0.000
The beta is: -87.451 and the delta is  0.972
The beta is: -35.723 and the delta is  0.514
The beta is: -57.037 and the delta is  0.765
The European Call Premium:  27.129


In [28]:
S = 120
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 3

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, callPayoff,  verbose = False)
print(f"The European Call Premium: {callPrc : 0.3f}")

The beta is: -93.742 and the delta is  1.000
The beta is: -93.742 and the delta is  1.000
The beta is: -7.641 and the delta is  0.126
The beta is: -92.500 and the delta is  1.000
The beta is: -43.805 and the delta is  0.605
The beta is: -63.735 and the delta is  0.821
The European Call Premium:  34.796


Delta increases over time

Problem #5

In [36]:
S = 100
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 1
n = 3

callPrc = americanBinomPricerRecursive(S, K, r, v, q, T, n, callPayoff,  verbose = False)
print(f"The American Call Premium: {callPrc : 0.3f}")

[56.64406241 33.8147424   0.          0.        ]
[56.64406241 15.04032855  0.          0.        ]
[56.64406241 15.04032855  0.          0.        ]
[33.14931753 15.04032855  0.          0.        ]
[33.14931753  6.6897296   0.          0.        ]
[18.28255221  6.6897296   0.          0.        ]
The American Call Premium:  18.283


Ther is no early exercise

In [34]:
S = 100
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 1
n = 3

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, callPayoff,  verbose = False)
print(f"The European Call Premium: {callPrc : 0.3f}")

The beta is: -92.500 and the delta is  1.000
The beta is: -62.627 and the delta is  0.773
The beta is:  0.000 and the delta is  0.000
The beta is: -74.186 and the delta is  0.903
The beta is: -27.688 and the delta is  0.444
The beta is: -47.516 and the delta is  0.708
The European Call Premium:  23.306


In [37]:
S = 100
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 1
n = 3

callPrc = americanBinomPricerRecursive(S, K, r, v, q, T, n, putPayoff,  verbose = False)
print(f"The American Put Premium: {callPrc : 0.3f}")

[ 0.          0.          3.89933488 30.57157332]
[ 0.          2.06235672  3.89933488 30.57157332]
[ 0.          2.06235672 20.40351726 30.57157332]
[ 1.09077967  2.06235672 20.40351726 30.57157332]
[ 1.09077967 11.7087201  20.40351726 30.57157332]
[ 6.67790123 11.7087201  20.40351726 30.57157332]
The American Put Premium:  6.678


In [38]:
S = 100
K = 95
r = 0.08
v = 0.30
q = 0.0
T = 1
n = 3

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, putPayoff,  verbose = False)
print(f"The European Put Premium: {callPrc : 0.3f}")

The beta is:  0.000 and the delta is  0.000
The beta is:  29.873 and the delta is -0.227
The beta is:  92.500 and the delta is -1.000
The beta is:  15.880 and the delta is -0.097
The beta is:  62.378 and the delta is -0.556
The beta is:  40.180 and the delta is -0.292
The European Put Premium:  11.002


The European Put is 11.002 and the American Put is 6.676. This means that there was some early exercises with the American option


Problem #6

In [39]:
S = 40
K = 40
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 3

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, putPayoff,  verbose = True)
print(f"The European Put Premium: {callPrc : 0.3f}")

The beta is:  0.000 and the delta is  0.000
The beta is:  17.241 and the delta is -0.323
The beta is:  39.470 and the delta is -1.000
The beta is:  9.751 and the delta is -0.146
The beta is:  29.584 and the delta is -0.629
The beta is:  20.839 and the delta is -0.365
[[40.   52.   67.6  87.88]
 [ 0.   32.   41.6  54.08]
 [ 0.    0.   25.6  33.28]
 [ 0.    0.    0.   20.48]]


[[ 6.247  2.149  0.     0.   ]
 [ 0.     9.445  3.801  0.   ]
 [ 0.     0.    13.87   6.72 ]
 [ 0.     0.     0.    19.52 ]]


[[-0.365 -0.146  0.     0.   ]
 [ 0.    -0.629 -0.323  0.   ]
 [ 0.     0.    -1.     0.   ]
 [ 0.     0.     0.     0.   ]]


[[20.839  9.751  0.     0.   ]
 [ 0.    29.584 17.241  0.   ]
 [ 0.     0.    39.47   0.   ]
 [ 0.     0.     0.     0.   ]]


The European Put Premium:  6.247


In [40]:
S = 40
K = 40
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 3

callPrc = euroBinomPricerRecursiveMatrix(S, K, r, v, q, T, n, callPayoff,  verbose = False)
print(f"The European Call Premium: {callPrc : 0.3f}")

The beta is: -39.470 and the delta is  1.000
The beta is: -22.230 and the delta is  0.677
The beta is:  0.000 and the delta is  0.000
The beta is: -29.197 and the delta is  0.854
The beta is: -9.363 and the delta is  0.371
The beta is: -17.593 and the delta is  0.635
The European Call Premium:  7.816


In [41]:
S = 40
K = 40
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 3

callPrc = americanBinomPricerRecursive(S, K, r, v, q, T, n, putPayoff,  verbose = False)
print(f"The American Put Premium: {callPrc : 0.3f}")

[ 0.          0.          3.16660277 11.16886174]
[ 0.          1.65788424  3.16660277 11.16886174]
[ 0.          1.65788424  7.84406148 11.16886174]
[ 0.86799019  1.65788424  7.84406148 11.16886174]
[ 0.86799019  4.8747172   7.84406148 11.16886174]
[ 2.95422756  4.8747172   7.84406148 11.16886174]
The American Put Premium:  2.954


In [42]:
S = 40
K = 40
r = 0.08
v = 0.30
q = 0.0
T = 0.5
n = 3

callPrc = americanBinomPricerRecursive(S, K, r, v, q, T, n, callPayoff,  verbose = False)
print(f"The American Call Premium: {callPrc : 0.3f}")

[13.01309727  7.05673217  0.          0.        ]
[13.01309727  3.26869392  0.          0.        ]
[13.01309727  3.26869392  0.          0.        ]
[7.73902974 3.26869392 0.         0.        ]
[7.73902974 1.51406624 0.         0.        ]
[4.37742951 1.51406624 0.         0.        ]
The American Call Premium:  4.377
